File: cpp/Classes/ferry/Ferry.cpp
//
// Created by dantezhu on 14-10-22.
//

#include "Ferry.h"
#include "netkit/Box.h"

namespace ferry {
Ferry *Ferry::getInstance() {
    // 用指针可以保证进程结束时，释放晚一些，不会报错
    static Ferry *instance;
    if (!instance) {
        instance = new Ferry();
    }
    return instance;
}

Ferry::Ferry() {
    pthread_mutex_init(&m_eventsMutex, NULL);
    pthread_mutex_init(&m_boxSnMutex, NULL);

    m_boxSn = 0;
}

Ferry::~Ferry() {
    stop();
    delAllCallbacks();
    clearEvents();

    pthread_mutex_destroy(&m_eventsMutex);
    pthread_mutex_destroy(&m_boxSnMutex);
}

ferry::Service*Ferry::getService() {
    return &m_service;
}

int Ferry::init(const std::string& host, short port) {
    return m_service.init(this, host, port);
}

void Ferry::start() {
    if (isRunning()) {
        return;
    }

    cocosSchedule();

    m_service.start();
}

void Ferry::stop() {
    if (!isRunning()) {
        return;
    }

    m_service.stop();

    cocosUnSchedule();
}

void Ferry::connect() {
    m_service.connect();
}

void Ferry::disconnect() {
    m_service.disconnect();
}

void Ferry::pauseSchedule() {
    cocos2d::Director::getInstance()->getScheduler()->pauseTarget(this);
}

void Ferry::resumeSchedule() {
    cocos2d::Director::getInstance()->getScheduler()->resumeTarget(this);
}

bool Ferry::isSchedulePaused() {
    return cocos2d::Director::getInstance()->getScheduler()->isTargetPaused(this);
}

bool Ferry::isConnected() {
    return m_service.isConnected();
}

bool Ferry::isRunning() {
    return m_service.isRunning();
}

void Ferry::delCallbacksForTarget(void *target) {
    delEventCallbacksForTarget(target);
    delRspCallbacksForTarget(target);
}

void Ferry::delAllCallbacks() {
    delAllEventCallbacks();
    delAllRspCallbacks();
}

void Ferry::send(netkit::IBox *box) {
    int sn = newBoxSn();
    setSnToBox(box, sn);

    m_service.send(box);
}

void Ferry::send(netkit::IBox *box, CallbackType callback, float timeout, void* target) {
    int sn = newBoxSn();
    setSnToBox(box, sn);

    struct timeval tvNow, tvTimeout;

    // 当前时间
    gettimeofday(&tvNow, NULL);

    // 超时
    tvTimeout.tv_sec = (int)(timeout);
    tvTimeout.tv_usec = (int)((timeout - tvTimeout.tv_sec) * 1000000);

    RspCallbackContainer *container = new RspCallbackContainer();
    container->sn = sn;

    timeradd(&tvNow, &tvTimeout, &container->expireTime);
    container->callback = callback;
    container->target = target;

    for (auto it = m_rspCallbacks.begin();; it ++) {
        if (it == m_rspCallbacks.end() ||
                timercmp(&(*it)->expireTime, &container->expireTime, >) ) {
            m_rspCallbacks.insert(it, container);
            break;
        }
    }

    m_service.send(box);
}

void Ferry::delRspCallbacksForTarget(void *target) {
    for(auto it = m_rspCallbacks.begin(); it != m_rspCallbacks.end();) {
        auto& container = *it;
        if (container->target == target) {
            // 先delete
            delete container;
            // 移除
            it = m_rspCallbacks.erase(it);
        }
        else {
            ++it;
        }
    }
}

void Ferry::delAllRspCallbacks() {
    for (auto& container: m_rspCallbacks) {
        delete container;
    }

    m_rspCallbacks.clear();
}

void Ferry::addEventCallback(CallbackType callback, void *target, const std::string &name) {
    EventCallbackContainer *container = new EventCallbackContainer();
    container->callback = callback;
    container->target = target;
    container->name = name;

    m_eventCallbacks.push_back(container);
}

void Ferry::delEventCallback(const std::string &name, void *target) {
    for(auto it = m_eventCallbacks.begin(); it != m_eventCallbacks.end();) {
        auto& container = *it;
        if (container->name == name && container->target == target) {
            delete container;
            // 移除
            it = m_eventCallbacks.erase(it);
        }
        else {
            ++it;
        }
    }
}

void Ferry::delEventCallbacksForTarget(void *target) {
    for(auto it = m_eventCallbacks.begin(); it != m_eventCallbacks.end();) {
        auto& container = *it;
        if (container->target == target) {
            delete container;
            // 移除
            it = m_eventCallbacks.erase(it);
        }
        else {
            ++it;
        }
    }
}

void Ferry::delAllEventCallbacks() {
    for(auto& container: m_eventCallbacks) {
        delete container;
    }

    m_eventCallbacks.clear();
}

void Ferry::onOpen(ferry::Service *service) {
    cocos2d::log("[%s][%d][%s]", __FILE__, __LINE__, __FUNCTION__);

    Event *event = new Event();
    event->what = EVENT_OPEN;
    postEvent(event);
}

void Ferry::onSend(ferry::Service *service, netkit::IBox *ibox) {
    cocos2d::log("[%s][%d][%s] box: %s", __FILE__, __LINE__, __FUNCTION__, ibox->toString().c_str());
    Event *event = new Event();
    event->what = EVENT_SEND;
    event->box = (netkit::Box*)ibox;
    postEvent(event);
}

void Ferry::onRecv(ferry::Service *service, netkit::IBox *ibox) {
    cocos2d::log("[%s][%d][%s] box: %s", __FILE__, __LINE__, __FUNCTION__, ibox->toString().c_str());

    Event *event = new Event();
    event->what = EVENT_RECV;
    event->box = (netkit::Box*)ibox;
    postEvent(event);
}

void Ferry::onClose(ferry::Service *service) {
    cocos2d::log("[%s][%d][%s]", __FILE__, __LINE__, __FUNCTION__);

    Event *event = new Event();
    event->what = EVENT_CLOSE;
    postEvent(event);
}

void Ferry::onError(ferry::Service *service, int code, netkit::IBox *ibox) {
    cocos2d::log("[%s][%d][%s] code: %d, box: %s",
        __FILE__, __LINE__, __FUNCTION__, code, ibox ? ibox->toString().c_str() : "null");

    Event *event = new Event();
    event->what = EVENT_ERROR;
    event->code = code;
    event->box = ibox;
    postEvent(event);
}

netkit::IBox* Ferry::createBox() {
    return new netkit::Box();
}

void Ferry::releaseBox(netkit::IBox* ibox) {
    if (ibox) {
        delete ibox;
    }
}

void Ferry::setSnToBox(netkit::IBox* ibox, int sn) {
    netkit::Box* box = (netkit::Box*)ibox;
    box->sn = sn;
}

int Ferry::getSnFromBox(netkit::IBox* ibox) {
    netkit::Box* box = (netkit::Box*)ibox;
    return box->sn;
}

void Ferry::loopEvents() {
    pthread_mutex_lock(&m_eventsMutex);
    // 防止访问冲突
    std::list<Event*> events;
    // 相当于直接剪切，速度快很多
    events.splice(events.end(), m_events);
    pthread_mutex_unlock(&m_eventsMutex);

    for (auto& event: events) {
        onEvent(event);
        delete event;
    }
}

void Ferry::postEvent(Event *event) {
    pthread_mutex_lock(&m_eventsMutex);

    m_events.push_back(event);

    pthread_mutex_unlock(&m_eventsMutex);
}

void Ferry::clearEvents() {
    pthread_mutex_lock(&m_eventsMutex);

    for (auto& event: m_events) {
        delete event;
    }

    m_events.clear();

    pthread_mutex_unlock(&m_eventsMutex);
}

void Ferry::onEvent(Event *event) {
    if (event->box && getSnFromBox(event->box) > 0) {
        // SEND, RECV, ERROR
        handleWithRspCallbacks(event);
    }
    else {
        // eventCallback只接受sn=0的，至于跨界面通信，这个由应用自己走eventbus
        handleWithEventCallbacks(event);
    }
}

int Ferry::newBoxSn() {
    pthread_mutex_lock(&m_boxSnMutex);

    m_boxSn %= 2100000000;

    // 即最小也是1
    int boxSn = ++m_boxSn;

    pthread_mutex_unlock(&m_boxSnMutex);

    return boxSn;
}

void Ferry::cocosSchedule() {
    auto func = [this](float dt){
        loopEvents();

        checkRspTimeout();
    };

    // 先调用这个
    cocos2d::Director::getInstance()->getScheduler()->schedule(func, this, 0, false, __FUNCTION__);
}

void Ferry::checkRspTimeout() {
    if (m_rspCallbacks.empty()) {
        return;
    }

    struct timeval tvNow;
    gettimeofday(&tvNow, NULL);

    std::list<RspCallbackContainer*> todoCallbacks;

    // 提前申请好，免得每次都要传
    // 因为纯删除，且挡住主线程，不会导致结构被破坏
    for(auto& container: m_rspCallbacks) {
        if (timercmp(&tvNow, &container->expireTime, >)) {
            todoCallbacks.push_back(container);
        }
        else {
            // 找到第一个没超时的，那么后面就都没有超时了
            break;
        }
    }

    if (todoCallbacks.empty()) {
        // 无需处理
        return;
    }

    // 再进入下一帧之前，不会释放
    Event *event = new Event();
    event->what = EVENT_TIMEOUT;

    for (auto& container: todoCallbacks) {
        // 没找到
        if (std::find(m_rspCallbacks.begin(), m_rspCallbacks.end(), container) == m_rspCallbacks.end()) {
            continue;
        }

        container->callback(event);

        delete container;

        m_rspCallbacks.remove(container);
    }

    delete event;
}

void Ferry::handleWithRspCallbacks(Event *event) {
    auto rspCallbacks = m_rspCallbacks;

    int sn = getSnFromBox(event->box);

    for(auto it = rspCallbacks.begin(); it != rspCallbacks.end(); ++it) {
        auto& container = *it;

        if (std::find(m_rspCallbacks.begin(), m_rspCallbacks.end(), container) == m_rspCallbacks.end()) {
            continue;
        }

        if (container->sn == sn)
        {
            // 调用
            container->callback(event);
            // 移除
            if (event->what == EVENT_RECV || event->what == EVENT_ERROR) {
                delete container;
                m_rspCallbacks.remove(container);
            }
        }
    }
}

void Ferry::handleWithEventCallbacks(Event *event) {
    auto eventCallbacks = m_eventCallbacks;

    for(auto it = eventCallbacks.begin(); it != eventCallbacks.end(); ++it) {
        auto& container = *it;
        if (std::find(m_eventCallbacks.begin(), m_eventCallbacks.end(), container) == m_eventCallbacks.end()) {
            // 已经中途被释放了
            continue;
        }

        container->callback(event);
    }
}

void Ferry::cocosUnSchedule() {
    cocos2d::Director::getInstance()->getScheduler()->unscheduleAllForTarget(this);
}
}

File: java/src/com/danniu/ferry/Ferry.java
package com.danniu.ferry;

import android.os.Handler;
import com.danniu.netkit.Box;
import com.danniu.netkit.IBox;
import de.greenrobot.event.EventBus;

import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Created by dantezhu on 14-12-16.
 */
public class Ferry implements Delegate {
    private Service service;
    private EventBus eventBus;
    private int boxSn;
    private Lock boxSnLock;
    private List<CallbackListener> rspCallbacks;
    private List<CallbackListener> eventCallbacks;
    private Handler timeoutChecker;

    public Ferry() {
        service = new Service();
        eventBus = new EventBus();
        eventBus.register(this);
        boxSnLock = new ReentrantLock();
        rspCallbacks = new LinkedList<CallbackListener>();
        eventCallbacks = new LinkedList<CallbackListener>();
        timeoutChecker = new Handler();
    }

    public int init(String host, int port) {
        // 准备好
        initTimeoutChecker();
        return service.init(this, host, port);
    }

    public void start() {
        if (isRunning()) {
            return;
        }

        service.start();
    }

    public void stop() {
        if (!isRunning()) {
            return;
        }

        service.stop();
    }

    public void connect() {
        service.connect();
    }

    public void disconnect() {
        service.disconnect();
    }

    public boolean isConnected() {
        return service.isConnected();
    }

    public boolean isRunning() {
        return service.isRunning();
    }

    public void delCallbacksForTarget(Object target) {
        delRspCallbacksForTarget(target);
        delEventCallbacksForTarget(target);
    }

    public void delAllCallbacks() {
        delAllRspCallbacks();
        delAllEventCallbacks();
    }

    public void send(IBox ibox) {
        int sn = newBoxSn();
        setSnToBox(ibox, sn);

        service.send(ibox);
    }

    public void send(IBox ibox, CallbackListener listener, float timeout, Object target) {
        int sn = newBoxSn();
        setSnToBox(ibox, sn);

        listener.target = target;
        listener.expireTime = (long)(timeout * 1000) + System.currentTimeMillis();
        listener.sn = sn;

        for (int i=0; i<rspCallbacks.size(); i++) {
            if (i == rspCallbacks.size()-1 ||
                    listener.expireTime < rspCallbacks.get(i).expireTime) {
                // 插入到i前面
                rspCallbacks.add(i, listener);
                break;
            }
        }
    }

    public void delRspCallbacksForTarget(Object target) {
        Iterator<CallbackListener> iterator = rspCallbacks.iterator();
        while (iterator.hasNext()) {
            CallbackListener listener = iterator.next();
            if (listener.target == target) {
                // 这种删除是安全的
                iterator.remove();
            }
        }
    }

    public void delAllRspCallbacks() {
        rspCallbacks.clear();
    }

    public void addEventCallback(CallbackListener listener, Object target, String name) {
        listener.target = target;
        listener.name = name;

        eventCallbacks.add(listener);
    }

    public void delEventCallback(String name, Object target) {
        Iterator<CallbackListener> iterator = eventCallbacks.iterator();
        while (iterator.hasNext()) {
            CallbackListener listener = iterator.next();
            if (listener.name.equals(name) && listener.target == target) {
                // 这种删除是安全的
                iterator.remove();
            }
        }
    }

    public void delEventCallbacksForTarget(Object target) {
        Iterator<CallbackListener> iterator = eventCallbacks.iterator();
        while (iterator.hasNext()) {
            CallbackListener listener = iterator.next();
            if (listener.target == target) {
                // 这种删除是安全的
                iterator.remove();
            }
        }
    }

    public void delAllEventCallbacks() {
        eventCallbacks.clear();
    }

    public void onOpen(Service service) {
        eventBus.post(new Event(Constants.EVENT_OPEN));
    }

    public void onSend(Service service, IBox ibox) {
        Event event = new Event(Constants.EVENT_SEND);
        event.box = ibox;

        eventBus.post(event);
    }

    public void onRecv(Service service, IBox ibox) {
        Event event = new Event(Constants.EVENT_RECV);
        event.box = ibox;

        eventBus.post(event);
    }

    public void onClose(Service service) {
        eventBus.post(new Event(Constants.EVENT_CLOSE));
    }

    public void onError(Service service, int code, IBox ibox) {
        Event event = new Event(Constants.EVENT_ERROR);
        event.code = code;
        event.box = ibox;

        eventBus.post(event);
    }

    public IBox createBox() {
        return new Box();
    }

    public void onEventMainThread(Event event) {
        if (event.box != null && getSnFromBox(event.box) > 0) {
            handleWithRspCallbacks(event);
        }
        else {
            handleWithEventCallbacks(event);
        }
    }

    private void handleWithRspCallbacks(Event event) {
        int sn = getSnFromBox(event.box);
        LinkedList<CallbackListener> tmpListeners = new LinkedList<CallbackListener>(rspCallbacks);

        for (CallbackListener listener: tmpListeners) {
            if (rspCallbacks.indexOf(listener) < 0) {
                // 没找到，说明被删除了
                continue;
            }

            if (listener.sn == sn) {
                callListener(listener, event);

                rspCallbacks.remove(listener);
            }
        }
    }

    private void handleWithEventCallbacks(Event event) {
        LinkedList<CallbackListener> tmpListeners = new LinkedList<CallbackListener>(eventCallbacks);

        for (CallbackListener listener: tmpListeners) {
            if (eventCallbacks.indexOf(listener) < 0) {
                // 没找到，说明被删除了
                continue;
            }

            callListener(listener, event);
        }
    }

    private void callListener(CallbackListener listener, Event event) {
        switch (event.what) {
            case Constants.EVENT_OPEN:
                listener.onOpen();
                break;
            case Constants.EVENT_CLOSE:
                listener.onClose();
                break;
            case Constants.EVENT_SEND:
                listener.onSend(event.box);
                break;
            case Constants.EVENT_RECV:
                listener.onRecv(event.box);
                break;
            case Constants.EVENT_ERROR:
                listener.onError(event.code, event.box);
                break;
            case Constants.EVENT_TIMEOUT:
                listener.onTimeout();
                break;
        }
    }

    private int newBoxSn() {
        boxSn %= 2100000000;

        boxSnLock.lock();
        try {
            return ++boxSn;
        }
        finally {
            boxSnLock.unlock();
        }
    }

    private void setSnToBox(IBox ibox, int sn) {
        Box box = (Box)ibox;
        box.sn = sn;
    }

    private int getSnFromBox(IBox ibox) {
        return ((Box)ibox).sn;
    }

    private void initTimeoutChecker() {
        timeoutChecker.postDelayed(new Runnable() {
            @Override
            public void run() {
                checkRspTimeout();
                timeoutChecker.postDelayed(this, Constants.TIMEOUT_CHECK_INTERVAL_MS);
            }
        }, Constants.TIMEOUT_CHECK_INTERVAL_MS);
    }

    private void checkRspTimeout() {
        if (rspCallbacks.isEmpty()) {
            return;
        }

        long nowTime = System.currentTimeMillis();

        LinkedList<CallbackListener> todoListeners = new LinkedList<CallbackListener>();

        for(CallbackListener listener: rspCallbacks) {
            if (listener.expireTime < nowTime) {
                // 说明超时了
                todoListeners.add(listener);
            }
            else {
                break;
            }
        }

        if (todoListeners.isEmpty()) {
            return;
        }

        Event event = new Event(Constants.EVENT_TIMEOUT);

        for (CallbackListener listener: todoListeners) {
            callListener(listener, event);
            rspCallbacks.remove(listener);
        }
    }

    public class CallbackListener {
        public String name;
        public Object target;
        public int sn;
        public long expireTime;

        public void onOpen() {}

        public void onSend(IBox ibox) {}

        public void onRecv(IBox ibox) {}

        public void onClose() {}

        public void onError(int code, IBox ibox) {}

        public void onTimeout() {}
    }

    public class Event {
        public int what;
        public IBox box;
        public int code;

        Event(int what_) {
            what = what_;
        }
    }
}
